### PR TITLE
Add order readiness info to status page

### PR DESCRIPTION
**Pull request by jknapp25**
Now we'll show:
- `readyBy` time before received
- "estimated ready time" after it has been marked received

image

### CODE REVIEW

File: packages/sdk/src/types.ts [hunk 1]
```diff
@@ -428,6 +428,8 @@ export type OrderFulfillment = {
   state: FulfillmentState;
   trackingNumber?: Maybe<Scalars['String']>;
   type: FulfillmentType;
+  minutesUntilReady?: Maybe<Scalars['Int']>;
+  createdAt?: Scalars['AWSDateTime'];
```
Comment (marisbest2) [inline packages/sdk/src/types.ts:432]: This isn't optional right?

File: packages/sdk/src/types.ts [hunk 2]
```diff
@@ -428,6 +428,8 @@ export type OrderFulfillment = {
   state: FulfillmentState;
   trackingNumber?: Maybe<Scalars['String']>;
   type: FulfillmentType;
+  minutesUntilReady?: Maybe<Scalars['Int']>;
```
Comment (marisbest2) [inline packages/sdk/src/types.ts:431]: I think you need readyBy etc here too right?

File: apps/patient/src/views/Status.tsx [hunk 3]
```diff
@@ -15,6 +26,67 @@ import { formatAddress, getFulfillmentType, preparePharmacy } from '../utils/gen
 import { orderStateMapping as m, text as t } from '../utils/text';
 import { useOrderContext } from './Main';
 import { HorizontalStatusStepper } from '../components/HorizontalStatusStepper';
+import dayjs from 'dayjs';
+
+function formatReadyTime(baseTime: string, minutesUntilReady: number): ReactElement {
```
Comment (marisbest2) [inline apps/patient/src/views/Status.tsx:31]: This should be a react component not a function. 

Use like 
```
<ReadyForPickup baseTime={createdAt} minutesUntilReady={minutesUntilReady} />
```

As currently set up, it'll rerender every time, whereas if its a component, React will only render it when props change.

File: apps/patient/src/views/Status.tsx [hunk 4]
```diff
@@ -15,6 +26,67 @@ import { formatAddress, getFulfillmentType, preparePharmacy } from '../utils/gen
...
+    <Text>
+      Ready{' '}
+      <b>
+        {dayPrefix} at {readyTime.format(timeFormat)}
+      </b>
+    </Text>
+  );
+}
+
+function formatReadyBy(readyBy: string, readyByDay: string): ReactElement {
```
Comment (marisbest2) [inline apps/patient/src/views/Status.tsx:59]: Same here, should be a component

File: apps/patient/src/views/Status.tsx (hunk 3 above)
Comment (marisbest2) [inline apps/patient/src/views/Status.tsx:31]: The naming here really is bad. This returns the **confirmed** ready time. But we have a different field called `readyByTime` which is the **desired** pickup time.

We should fix the naming on both ends probs, but in the meantime this should probably be `<ReadyForPickupTime>` or something like that.

File: apps/patient/src/graphql/queries.ts [hunk 5]
```diff
@@ -6,6 +6,9 @@ export const GET_ORDER = gql`
       id
       state
       isReroutable
+      readyBy
+      readyByDay
+      readyByTime
```
Comment (marisbest2) [inline apps/patient/src/graphql/queries.ts:11]: Are we using `readyByTime`?

File: apps/patient/src/graphql/queries.ts [hunk 6]
```diff
@@ -22,6 +25,8 @@ export const GET_ORDER = gql`
         type
         state
         trackingNumber
+        minutesUntilReady
+        createdAt
```
Comment (marisbest2) [inline apps/patient/src/graphql/queries.ts:29]: Is `minutesUntilReady` actually relative to `createdAt`? It looks to me from reading the code that `minutesUntilReady` is relative to that being set whereas `createdAt` happens when the order is sent?

**Review by marisbest2 — CHANGES_REQUESTED** [review]
A few changes and also a big question - Is `minutesUntilReady` actually relative to `createdAt`?

File: packages/sdk/src/types.ts (hunk 2 above)
Comment (jknapp25) [reply packages/sdk/src/types.ts:431]: it's on the original order already

File: apps/patient/src/graphql/queries.ts (hunk 5 above)
Comment (jknapp25) [reply apps/patient/src/graphql/queries.ts]: no, good catch

File: apps/patient/src/views/Status.tsx (hunk 3 above)
Comment (marisbest2) [reply apps/patient/src/views/Status.tsx:31]: pretty much, as a function, this will rerender every time because it won't be part of the shadow DOM. The code isn't bad performing, but its not ideal.

If its a component, the code WILL be part of the shadow DOM, so it'll only rerender when the parent rerenders (never) or the props change (also never).

File: apps/patient/src/views/Status.tsx [hunk 7]
```diff
@@ -15,6 +26,67 @@ import { formatAddress, getFulfillmentType, preparePharmacy } from '../utils/gen
 import { orderStateMapping as m, text as t } from '../utils/text';
 import { useOrderContext } from './Main';
 import { HorizontalStatusStepper } from '../components/HorizontalStatusStepper';
+import dayjs from 'dayjs';
+
+function formatReadyTime(baseTime: string, minutesUntilReady: number): ReactElement {
+  const currentTime = dayjs(baseTime);
```
Comment (marisbest2) [inline apps/patient/src/views/Status.tsx:32]: This naming is confusing to me also. I keep reading `currentTime` as "now" when its really the `baseTime` just as a `dayjs`.

Comment (jknapp25) [reply apps/patient/src/views/Status.tsx:32]: yeah this went thru a few revisions, goods points, I'll address

File: apps/patient/src/components/ReadyText.tsx [hunk 8]
```diff
@@ -0,0 +1,106 @@
+import dayjs, { Dayjs } from 'dayjs';
+import { Text } from '@chakra-ui/react';
+import { OrderFulfillment } from 'packages/sdk/dist/types';
+
+export type Maybe<T> = T | null;
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:5]: Presumably this exists elsewhere (e.g. in `packages/sdk/dist/types`). `Maybe<T>` should also be `T | null | undefined`

File: apps/patient/src/components/ReadyText.tsx [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    <>
+      {showPatientDesiredreadyBy && (
+        <PatientDesiredReadyBy readyBy={readyBy} readyByDay={readyByDay} />
+      )}
+      {showEstimatedReadyAt && (
+        <PharmacyEstimatedReadyAt pharmacyEstimatedReadyAt={fulfillment.pharmacyEstimatedReadyAt} />
+      )}
+    </>
+  );
+};
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:35]: Its hard to parse this. I _think_ this is:
- If deliveryPharmacy = do nothing
- If SENT = show desired
- If RECEIVED = show estimated ready

Is that right?
If yes, I think we can simplify this to:

```suggestion
  if (isDeliveryPharmacy) return null;
  
  if ((!fulfillment || fulfillment?.state === 'SENT') && readyBy && readyByDay) {
    return <PatientDesiredReadyBy readyBy={readyBy} readyByDay={readyByDay} />
  }
  
  if (fulfillment?.state === 'RECEIVED' && fulfillment?.pharmacyEstimatedReadyAt) {
    return <PharmacyEstimatedReadyAt pharmacyEstimatedReadyAt={fulfillment.pharmacyEstimatedReadyAt} />
  }
```

This makes it clearer that only one of these is being returned and when.

File: apps/patient/src/components/ReadyText.tsx [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+const roundUpTo15MinInterval = (pharmacyEstimatedReadyAt: Date): Dayjs => {
+  const dayJS = dayjs(pharmacyEstimatedReadyAt);
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:38]:
```suggestion
  const estimatedReadyAtAsDayJs = dayjs(pharmacyEstimatedReadyAt);
```

`dayJS` and `dayjs` are VERY close to each other

File: apps/patient/src/components/ReadyText.tsx [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    return dayJS.minute(roundedMinutes);
+  }
+};
+
+interface PharmacyEstimatedReadyAtProps {
+  pharmacyEstimatedReadyAt: Date;
+}
+const PharmacyEstimatedReadyAt = ({ pharmacyEstimatedReadyAt }: PharmacyEstimatedReadyAtProps) => {
+  const readyAt = roundUpTo15MinInterval(pharmacyEstimatedReadyAt);
+  const isTomorrow = readyAt.isSame(dayjs().add(1, 'day'), 'day');
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:53]: Consider: https://day.js.org/docs/en/plugin/is-tomorrow

File: apps/patient/src/components/ReadyText.tsx [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    <Text>
+      Ready{' '}
+      {isTomorrow ? (
+        <b>Tomorrow at {readyAt.format(timeFormat)}</b>
+      ) : (
+        <>
+          at <b>{readyAt.format(timeFormat)}</b>
+        </>
+      )}
+    </Text>
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:66]:
```suggestion
    <Text>
      Ready{isTomorrow ? <b> Tomorrow </b> : " "} at <b>{readyAt.format(timeFormat)}</b>
    </Text>
```

Something like this maybe cleaner. Might need to ensure the spaces are correct.

File: apps/patient/src/components/ReadyText.tsx [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+}
+const PatientDesiredReadyBy = ({ readyBy, readyByDay }: PatientDesiredReadyByProps) => {
+  if (readyBy === 'Urgent') {
+    return (
+      <Text>
+        Need order <b>as soon as possible</b>
+      </Text>
+    );
+  } else if (readyBy === 'After hours') {
+    return readyByDay === 'Today' ? (
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:82]: What happens if they said they need it tomorrow, but now its tomorrow? i.e. on Monday they said tomorrow and now its tuesday.

File: apps/patient/src/components/ReadyText.tsx [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        <b>
+          {hour} {period}
+          {readyByDay === 'Tomorrow' ? ' tomorrow' : ''}
+        </b>
+      </Text>
+    );
+  }
+};
+
+export default ReadyText;
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:106]: Generally, prefer named exports (unless theres a reason that the library we're using doesnt like that, but I dont think thats the case here)

File: packages/sdk/src/types.ts (hunk 1 above)
Comment (marisbest2) [reply packages/sdk/src/types.ts:433]: This still says optional

File: apps/patient/src/components/ReadyText.tsx (hunk 9 above)
Comment (jknapp25) [reply apps/patient/src/components/ReadyText.tsx:35]: oh nice, way better

File: apps/patient/src/components/ReadyText.tsx (hunk 8 above)
Comment (jknapp25) [reply apps/patient/src/components/ReadyText.tsx:5]: imported

File: apps/patient/src/components/ReadyText.tsx (hunk 13 above)
Comment (jknapp25) [reply apps/patient/src/components/ReadyText.tsx:82]: oh good catch

File: apps/patient/src/components/ReadyText.tsx [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+import isTomorrow from 'dayjs/plugin/isTomorrow';
+
+dayjs.extend(isTomorrow);
+
+interface ReadyTextProps {
+  readyBy?: string;
+  readyByDay?: string;
+  readyByTime?: string;
+  isDeliveryPharmacy?: boolean;
+  fulfillment?: Maybe<OrderFulfillment> | undefined;
```
Comment (marisbest2) [inline apps/patient/src/components/ReadyText.tsx:13]:
```suggestion
  fulfillment?: Maybe<OrderFulfillment>;
```

**Review by marisbest2 — APPROVED** [review]

File: apps/patient/src/components/ReadyText.tsx (hunk 15 above)
Comment (jknapp25) [reply apps/patient/src/components/ReadyText.tsx]: cool
